### PR TITLE
[ECDH API change] Support custom hash function

### PR DESCRIPTION
**Pull request by fanatid**
Solve #352

### CODE REVIEW

**Comment by fanatid** [comment]
I think that `unsafeECDH` that return public key as result will be better than `ecdh` with custom hash function, so I close this PR.

**Comment by fanatid** [comment]
Rebased. @sipa now you have 2 PRs, hash function as argument (this) and raw function (#446).. please select one of them, review and merge :)

**Comment by chfast** [comment]
Yes please :) For me it does not matter which variant it will be, but I need access to raw point coordinates (precisely only x).

Actually, this variant is a bit better than mine, because the serialized point is not useful for me directly.

File: src/modules/ecdh/tests_impl.h [hunk 1]
```diff
@@ -7,6 +7,22 @@
 #ifndef _SECP256K1_MODULE_ECDH_TESTS_
 #define _SECP256K1_MODULE_ECDH_TESTS_
 
+int ecdh_hash_function_test_fail(unsigned char *output, const unsigned char *x, const unsigned char *y) {
+    if (1) {
+        return 0;
+    }
```
Comment (dcousens) [inline src/modules/ecdh/tests_impl.h]: is this meant to be here?

File: include/secp256k1_ecdh.h [hunk 2]
```diff
@@ -7,21 +7,41 @@
 extern "C" {
 # endif
 
+/** A pointer to a function that apply hash function to a point
```
Comment (sipa) [inline include/secp256k1_ecdh.h]: applies

Comment (fanatid) [inline include/secp256k1_ecdh.h]: fixed

File: src/modules/ecdh/tests_impl.h (hunk 1 above)
Comment (fanatid) [inline src/modules/ecdh/tests_impl.h]: removed

**Comment by chfast** [comment]
How is this going on?

**Comment by chfast** [comment]
Ping.

**Comment by apoelstra** [comment]
I'll rebase and review it.

**Comment by apoelstra** [comment]
@fanatid I'm not able to rebase your PR it seems. Can you do it? If you want, my version of the commit is at https://github.com/apoelstra/secp256k1/tree/custom-ecdh-hash-function

**Comment by fanatid** [comment]
@apoelstra rebased

File: include/secp256k1_ecdh.h [hunk 3]
```diff
@@ -7,21 +7,41 @@
...
 # endif
 
+/** A pointer to a function that applies hash function to a point
+ *
+ *  Returns: 1 if a point was successfully hashed. 0 will cause ecdh to fail
+ *  Out:    output:     pointer to an array to be filled by the function
+ *  In:     x:          pointer to a 32-byte x coordinate
+ *          y:          pointer to a 32-byte y coordinate
+ */
+typedef int (*secp256k1_ecdh_hash_function)(
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply include/secp256k1_ecdh.h]: I think it's reasonable. Imagine a hash function that returns the x coordinate, but only if y is odd (or has Jacobi symbol 1, or whatever the parity is).

File: include/secp256k1_ecdh.h [hunk 4]
```diff
@@ -7,21 +7,41 @@
...
  *  Returns: 1: exponentiation was successful
  *           0: scalar was invalid (zero or overflow)
  *  Args:    ctx:        pointer to a context object (cannot be NULL)
- *  Out:     result:     a 32-byte array which will be populated by an ECDH
- *                       secret computed from the point and scalar
+ *  Out:     output:     pointer to an array to be filled by the function
  *  In:      pubkey:     a pointer to a secp256k1_pubkey containing an
  *                       initialized public key
  *           privkey:    a 32-byte scalar with which to multiply the point
+ *           hashfp:		 pointer to a hash function. If NULL, secp256k1_ecdh_hash_function_sha256 is used
```
Comment (apoelstra) [inline include/secp256k1_ecdh.h]: awkward spacing

**Comment by apoelstra** [comment]
ACK aside from spacing nit

File: include/secp256k1_ecdh.h (hunk 4 above)
Comment (fanatid) [reply include/secp256k1_ecdh.h]: fixed

**Comment by apoelstra** [comment]
cc @sipa

**Comment by sipa** [comment]
Hmm, is there no need for a data parameter that gets passed to the hashing function? This way the hash function cannot access any application dependent data, unless through a global (which is not threadsafe)?

**Comment by chfast** [comment]
> Hmm, is there no need for a data parameter that gets passed to the hashing function? This way the hash function cannot access any application dependent data, unless through a global (which is not threadsafe)?

Is this a recommended change in this PR?

**Comment by chfast** [comment]
I can add it then if @fanatid does not have time.

**Comment by fanatid** [comment]
Pointer to arbitrary data added to hash function.

**Comment by ofek** [comment]
Is this ready to be merged?

**Comment by sipa** [comment]
utACK c8fbc3c397b547bc64435a9bffb8f989cd23aba0

**Comment by apoelstra** [comment]
ACK

**Comment by chfast** [comment]
Can this be merged finally?

**Comment by chfast** [comment]
> Can this be merged finally?

Can it?

**Pull request merged**